<launch>

  <arg name="image_topic" default="/camera/image_raw" />
  <!-- Vision arguments -->
  <arg name="scale" default="1.0" />
  <arg name="rate" default="4" />
  <arg name="hue" default="true" />
  <arg name="intensity" default="true" />
  <arg name="texton" default="true" />
  <arg name="orb" default="true" />
  <arg name="surf" default="false" />
  <arg name="use_pc" default="false" />
  <arg name="use_tf" default="false" />
  <arg name="depth_cloud_topic" default="" />
  <arg name="transform_topic" default="" />
  
  <!-- Topic model arguments -->
  <arg name="K" default="100" />
  <arg name="V" default="16180" />
  <arg name="alpha" default="0.1" />
  <arg name="beta" default="1.0" />
  <arg name="gamma" default="0" />
  <arg name="tau" default="0.5" />
  <arg name="p_refine_last_observation" default="0.5" />
  <arg name="num_threads" default="4" />
  <arg name="cell_space" default="32" />
  <arg name="cell_time" default="3600" />
  <arg name="G_time" default="1" />
  <arg name="G_space" default="1" />
  <arg name="polled_refine" default="false" />
  <arg name="words_topic" default="/rost_vision/words" />
  <arg name="min_obs_refine_time" default="200" />
  <arg name="word_obs_queue_size" default="1" />
  <arg name="map_ppx_type" default="global" />

  <!-- Visualization arguments -->
  <!--<arg name="words_topic" default="/rost_vision/words" />-->
  <arg name="map_topic" default="/rost/topic_map" />
  <arg name="ppx_topic" default="/ppx_cloud" />
  <arg name="map_type" default="TopicMap" />
  <arg name="show_topics" default="true" />
  <arg name="2d_words_topic" default="/rost_vision/words_2d" />
  <arg name="show_words" default="true" />
  <arg name="show_perplexity" default="true" />
  <arg name="show_equalized" default="true" />
  <arg name="ppx_display_factor" default="0.75" />

  <!-- Launch word extractor -->
  <node pkg="sunshine" type="visual_word_extractor" name="rost_vision">
    <!--param name="vocabulary" type="string" value="$(find summarizer)/share/vocab/underwater1k.yml"/-->
    <!--<param name="vocabulary" type="string" value="$(find rost_vision)/data/orb_vocab/barbados2010_5000.yml"/>-->
    <param name="feature_descriptor" type="string" value="ORB"/>
    <param name="scale" type="double" value="$(arg scale)"/>
    <param name="rate" type="double" value="$(arg rate)"/>
    <param name="use_hue" value="$(arg hue)"/>
    <param name="use_intensity" value="$(arg intensity)"/>
    <param name="use_orb" value="$(arg orb)"/>
    <param name="use_surf" value="$(arg surf)"/>
    <param name="use_texton" value="$(arg texton)" />
    <param name="image" value="$(arg image_topic)"/>
    <param name="transform" value="$(arg transform_topic)"/>
    <param name="pc" value="$(arg depth_cloud_topic)" />
    <param name="use_pc" value="$(arg use_pc)" />
    <param name="use_tf" value="$(arg use_tf)" />
    <param name="publish_2d_words" value="true" />
  </node>
  

  <!-- Launch topic model -->
  <node pkg="sunshine" type="topic_model" name="rost">
    <param name="K" type="int" value="$(arg K)" />
    <param name="V" type="int" value="$(arg V)" />
    <param name="alpha" type="double" value="$(arg alpha)" />
    <param name="beta" type="double" value="$(arg beta)" />
    <param name="gamma" type="double" value="$(arg gamma)" />
    <param name="tau" type="double" value="$(arg tau)" />
    <param name="p_refine_last_observation" type="double" value="$(arg p_refine_last_observation)" />
    <param name="num_threads" type="int" value="$(arg num_threads)" />
    <param name="cell_space" type="double" value="$(arg cell_space)" />
    <param name="cell_time" type="double" value="$(arg cell_time)" />
    <param name="G_time" type="int" value="$(arg G_time)" />
    <param name="G_space" type="int" value="$(arg G_space)" />
    <param name="polled_refine" value="$(arg polled_refine)" />
    <param name="words_topic" value="$(arg words_topic)" />
    <param name="min_obs_refine_time" value="$(arg min_obs_refine_time)" />
    <param name="word_obs_queue_size" value="$(arg word_obs_queue_size)" />
    <param name="publish_map" type="bool" value="true" />
    <param name="map_ppx_type" type="string" value="$(arg map_ppx_type)" />
  </node>

  <node pkg="sunshine" type="visualize" name="word_visualizer">
    <param name="ppx_topic" type="string" value="/rost/scene_perplexity" />
    <param name="words_topic" type="string" value="$(arg 2d_words_topic)" />
    <param name="topics_topic" type="string" value="/rost/topics" />
    <param name="show_topics" type="bool" value="$(arg show_topics)" />
    <param name="show_words" type="bool" value="$(arg show_words)" />
    <param name="show_perplexity" type="bool" value="$(arg show_perplexity)" />
    <param name="show_equalized" type="bool" value="$(arg show_equalized)" />
   </node>
  
  <node type="rviz" name="rviz" pkg="rviz" args="-d $(find sunshine)/rviz/imwalker.rviz" />
</launch>
